### ISSUE TITLE
`t.test` resolves before async teardowns complete

### ISSUE THREAD

**Issue by coreyfarrell** [tapjs/tapjs#644]
IMO the following should pass:
```js
import {promisify} from 'util';
import t from 'tap';

const delay = promisify(setTimeout);

t.test('test', async t => {
  let finished = false;
  await t.test('subtest', async t => {
    t.teardown(async () => {
      await delay(50);
      finished = true;
    });
  });
  t.equal(finished, true);
});
```

It's a subtle difference but I think it's important.  I'll try looking into this if I'm able to set aside some time soon but I wanted to post an issue so it's not forgotten.

**Comment by psixdev** [comment]
Hello. I'm trying to figure out how tap works and stumbled upon an interesting problem: several `teardown` run in parallel (unlike beforeEach and afterEach). I think these problems can be related. If not, I can start a separate issue.

Here is a test example:
```js
const t = require('tap');
const {promisify} = require('util');

const delay = promisify(setTimeout);

t.beforeEach(async () => {
	console.error('beforeEach 1: start');
	await delay(50);
	console.error('beforeEach 1: finish');
});

t.beforeEach(async () => {
	console.error('beforeEach 2: start');
	await delay(50);
	console.error('beforeEach 2: finish');
});

t.afterEach(async () => {
	console.error('afterEach 1: start');
	await delay(50);
	console.error('afterEach 1: finish');
});

t.afterEach(async () => {
	console.error('afterEach 2: start');
	await delay(50);
	console.error('afterEach 2: finish');
});

t.teardown(async () => {
	console.error('teardown 1: start');
	await delay(50);
	console.error('teardown 1: finish');
});

t.teardown(async () => {
	console.error('teardown 2: start');
	await delay(50);
	console.error('teardown 2: finish');
});

t.test('test', async () => {
	console.error('test: start');
	await delay(50);
	console.error('test: finish');
});
```

And the output:
```
beforeEach 1: start
beforeEach 1: finish
beforeEach 2: start
beforeEach 2: finish
test: start
test: finish
afterEach 1: start
afterEach 1: finish
afterEach 2: start
afterEach 2: finish
teardown 1: start
teardown 2: start
teardown 1: finish
teardown 2: finish
```

Look at this:
https://github.com/tapjs/node-tap/blob/56fb7b239647bb57cd1f161040c4c5dad9eb1512/lib/test.js#L537

This line of code confuses me. Because of it, `emitSubTeardown` is called in `process`, and also `emitSubTeardown` is called below in `ret.then`. Judging by the code, sequential execution of `teardown` was expected.

**Comment by daniel-rodrigue** [comment]
> https://github.com/tapjs/node-tap/blob/56fb7b239647bb57cd1f161040c4c5dad9eb1512/lib/test.js#L537
> 
> This line of code confuses me. Because of it, `emitSubTeardown` is called in `process`, and also `emitSubTeardown` is called below in `ret.then`. Judging by the code, sequential execution of `teardown` was expected.

My understanding is it's missing some `this.processing = false` logic or something like that, otherwise the queue will continue to be processed, causing the next tearDown in the queue to be called before the first one is completed.

**Comment by isaacs** [comment]
This passes now in tap 18 (not sure if it was ever passing before, but it definitely does now).
